### PR TITLE
Add a pair of options to kcl consume for unmarshalling kafka messages in protobuf to json

### PR DESCRIPTION
**Pull request by moming00**
hello, it will be good if kcl supports unmarshalling kafka messages in protobuf format to Json.
So introduced two commandline options: --proto-file to specify the Proto Source File or ProtoSet File which includes the original message defination, --proto-message to specify the message name in "packageName.messageName" format. For example:
`kcl consume mytopic -n=1 --proto-file=/usr/local/src/proto_Generator/sample.proto --proto-message=sample.FileState`
`kcl consume mytopic -n=1 --proto-file=/usr/local/src/proto_Generator/sample.protoset --proto-message=sample.FileState`

this could return message like:
`{
        "Message": "Hello World",
        "Id": "7b1271413a48a80cfa0c38e83788b32a",
        "Timestamp": "2022-12-13T16:03:30.760+08:00",
        "HostName": "ck10",
        "Path": "/data01//flow.log",
        "RowNumer": "59",
        "Seq": "59",
        "TaskId": "111359604125616",
        "Topic": "fc111359604125096",
        "IP": "192.168.110.66"
}`

proto-file with "protoset" format supposed to be more efficient, as which already includs compiled info from source proto file, so requires no parsing in runtime. Aslo protoset file could be self-contained if `--include_imports` is specified when generating the file, while proto souce file may require additional efforts to locate the imported protos.

### CODE REVIEW

File: commands/consume/command.go [hunk 1]
```diff
@@ -30,6 +30,8 @@ func (c *consumption) command() *cobra.Command {
 	cmd.Flags().DurationVar(&c.fetchMaxWait, "fetch-max-wait", 5*time.Second, "maximum amount of time to wait when fetching from a broker before the broker replies")
 	cmd.Flags().StringVar(&c.rack, "rack", "", "the rack to use for fetch requests; setting this opts in to nearest replica fetching (Kafka 2.2.0+)")
 	cmd.Flags().BoolVar(&c.readUncommitted, "read-uncommitted", false, "opt in to reading uncommitted offsets")
+	cmd.Flags().StringVar(&c.protoFile, "proto-file", "", "proto file includes the definition of message structure")
```
Comment (twmb) [inline commands/consume/command.go:33]: `an optional proto definition file to decode protobuf messages, requires --proto-message`

File: commands/consume/command.go [hunk 2]
```diff
@@ -30,6 +30,8 @@ func (c *consumption) command() *cobra.Command {
...
+	cmd.Flags().StringVar(&c.protoMessage, "proto-message", "", "name of a protobuf message structure, should in 'package.message' format")
```
Comment (twmb) [inline commands/consume/command.go:34]: `the proto.message structure in --proto-file to use for decoding, requires --proto-file`

(is my phrasing accurate?)

File: commands/consume/protobuf_decoder.go [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+	"fmt"
+	"os"
+
+	"github.com/golang/protobuf/jsonpb"
+	"github.com/jhump/protoreflect/desc"
+	"github.com/jhump/protoreflect/desc/protoparse"
+	"github.com/jhump/protoreflect/dynamic"
+)
+
+type PBDecoder struct {
```
Comment (twmb) [inline commands/consume/protobuf_decoder.go:13]: can this be un-exported? `pbDecoder`

File: commands/consume/protobuf_decoder.go [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+	message string
+	md      *desc.MessageDescriptor
+}
+
+func NewPBDecoder(protoFile string, messageName string) (pbs *PBDecoder, err error) {
```
Comment (twmb) [inline commands/consume/protobuf_decoder.go:18]: This function and the function below can also be un-exported

File: commands/consume/protobuf_decoder.go [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+	}
+	return &PBDecoder{md: md, message: messageName}, nil
+}
+
+func (c *PBDecoder) JsonString(pbData []byte) (bytes []byte, err error) {
+	dyMsg := dynamic.NewMessage(c.md)
+	if err = dyMsg.Unmarshal(pbData); err != nil {
+		return nil, err
+	}
+	if bytes, err = dyMsg.MarshalJSONPB(&jsonpb.Marshaler{Indent: "\t"}); err != nil {
```
Comment (twmb) [inline commands/consume/protobuf_decoder.go:41]: Why tab indenting?

File: commands/consume/protobuf_decoder.go [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+	"os"
+
+	"github.com/golang/protobuf/jsonpb"
+	"github.com/jhump/protoreflect/desc"
+	"github.com/jhump/protoreflect/desc/protoparse"
+	"github.com/jhump/protoreflect/dynamic"
+)
+
+type PBDecoder struct {
+	message string
```
Comment (twmb) [inline commands/consume/protobuf_decoder.go:14]: I don't think `message` is needed in this struct, right?

File: commands/consume/command.go [hunk 7]
```diff
@@ -30,6 +30,8 @@ func (c *consumption) command() *cobra.Command {
...
 	return cmd
```
Comment (twmb) [inline commands/consume/command.go:35]: Can this also have `cmd.Flags().MarkFlagsRequiredTogether("proto-file", "proto-message")`?

https://pkg.go.dev/github.com/spf13/cobra#Command.MarkFlagsRequiredTogether

File: commands/consume/command.go (hunk 1 above)
Comment (moming00) [reply commands/consume/command.go:33]: updated

File: commands/consume/protobuf_decoder.go (hunk 3 above)
Comment (moming00) [reply commands/consume/protobuf_decoder.go:13]: sure

File: commands/consume/protobuf_decoder.go (hunk 5 above)
Comment (moming00) [reply commands/consume/protobuf_decoder.go:41]: trying to print a pretty formatted json string, pls let me know if you think it's no necessary

File: commands/consume/protobuf_decoder.go (hunk 6 above)
Comment (moming00) [reply commands/consume/protobuf_decoder.go:14]: true

File: commands/consume/command.go (hunk 7 above)
Comment (moming00) [reply commands/consume/command.go:35]: thx for the new knowledge, updated the version of flag package accordingly.

File: commands/consume/command.go (hunk 2 above)
Comment (moming00) [reply commands/consume/command.go:34]: yup, and I would like to keep the `should in 'PackageName.MessageName' format"` part, please let me know your thoughts.
Here is an example: `--proto-message=sample.FileState`

**Comment by moming00** [comment]
there is voice suggesting to use the compiled proto file, will reopen the PR after everything settled

**Comment by twmb** [comment]
fwiw compiled proto file might be difficult to use as well -- I usually work with the non-compiled files. Perhaps support both?

**Comment by moming00** [comment]
> fwiw compiled proto file might be difficult to use as well -- I usually work with the non-compiled files. Perhaps support both in the same flag?

this may be the solution =)

**Comment by twmb** [comment]
Awesome, lgtm, ty. I'll bump the latest franz pin and release.

**Pull request merged**